File: apps/desktop/src/features/todo/shared/activity-format.test.ts
/**
 * activity-format 单测：历史行文案格式化（changes 前后值 / 旧行 fields 回退 /
 * 标签动作 / 枚举值口径 / 长文本截断 / 非法 JSON 兜底）
 */
import { describe, expect, it } from "vitest";

import { describeActivity } from "./activity-format";

describe("describeActivity", () => {
  it("update 新行：changes 渲染前后值（优先级中文档位、null=无）", () => {
    const detail = JSON.stringify({
      fields: ["priority", "due_date"],
      changes: [
        { field: "priority", from: 0, to: 4 },
        { field: "due_date", from: null, to: "2026-09-25" },
      ],
    });
    expect(describeActivity("update", detail)).toBe(
      "更新（优先级：无 → 紧急、截止日期：无 → 2026-09-25）",
    );
  });

  it("update 旧行：无 changes 回退字段名清单", () => {
    const detail = JSON.stringify({ fields: ["priority", "project_id"] });
    expect(describeActivity("update", detail)).toBe("更新（优先级、所属项目）");
  });

  it("枚举/布尔/百分比值口径", () => {
    const detail = JSON.stringify({
      changes: [
        { field: "status", from: "pending", to: "doing" },
        { field: "is_favorite", from: 0, to: 1 },
        { field: "percent_done", from: 0, to: 60 },
        { field: "position", from: 1.23, to: 4.56 },
      ],
    });
    expect(describeActivity("update", detail)).toBe(
      "更新（状态：待办 → 进行中、收藏：否 → 是、进度：0% → 60%、顺序：已调整 → 已调整）",
    );
  });

  it("文本值超 30 字截断加省略号", () => {
    const long = "这是一段非常长的任务标题用于验证前端截断逻辑是否生效应该正好超过三十个字符";
    const detail = JSON.stringify({ changes: [{ field: "title", from: "短", to: long }] });
    const out = describeActivity("update", detail);
    expect(out).toContain(`${long.slice(0, 30)}…`);
    expect(out).not.toContain(long);
  });

  it("标签挂/摘动作拼标签名", () => {
    expect(describeActivity("label_add", JSON.stringify({ label: "工作" }))).toBe(
      "添加标签「工作」",
    );
    expect(describeActivity("label_remove", JSON.stringify({ label: "工作" }))).toBe(
      "移除标签「工作」",
    );
  });

  it("从属对象动作统一拼 target（子任务/关联/提醒）", () => {
    expect(describeActivity("subtask_add", JSON.stringify({ target: "写初稿" }))).toBe(
      "添加子任务「写初稿」",
    );
    expect(describeActivity("link_remove", JSON.stringify({ target: "对方任务" }))).toBe(
      "移除关联任务「对方任务」",
    );
    expect(
      describeActivity("reminder_add", JSON.stringify({ target: "2026-09-21 09:00" })),
    ).toBe("添加提醒「2026-09-21 09:00」");
    // target 缺失回退纯动作文案
    expect(describeActivity("comment_delete", "{}")).toBe("删除评论");
  });

  it("repeat_rule 伪字段：六字段快照对象走 repeatLabel 口径", () => {
    const snap = (mode: number, after: number, weekdays = 0) => ({
      mode,
      after,
      weekdays,
      end_type: 0,
      end_param: 0,
      from_done: 0,
    });
    const detail = JSON.stringify({
      changes: [{ field: "repeat_rule", from: snap(0, 0), to: snap(2, 1, 2) }],
    });
    expect(describeActivity("update", detail)).toBe("更新（重复规则：不重复 → 每周二）");
  });

  it("附件与子任务改名动作拼 target", () => {
    expect(describeActivity("attachment_add", JSON.stringify({ target: "报告.pdf" }))).toBe(
      "添加附件「报告.pdf」",
    );
    expect(describeActivity("attachment_delete", JSON.stringify({ target: "报告.pdf" }))).toBe(
      "移除附件「报告.pdf」",
    );
    expect(
      describeActivity("subtask_rename", JSON.stringify({ target: "旧名 → 新名" })),
    ).toBe("子任务改名「旧名 → 新名」");
  });

  it("非 update 动作与非法 JSON 回退动作文案", () => {
    expect(describeActivity("complete", "{}")).toBe("标记为完成");
    expect(describeActivity("create", "not-json")).toBe("创建了任务");
    expect(describeActivity("future_action", "{}")).toBe("future_action");
  });
});

File: apps/mobile/test/activity_format_test.dart
// describeActivity 纯函数测试：镜像桌面 activity-format.test.ts 口径
// （三代 detail 形状回退 + 字段值格式化 + 非法 JSON 兜底）
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:orbit/modules/todo/logic/activity_format.dart';

void main() {
  group('describeActivity', () {
    test('新行 changes：枚举走标签口径、字符串原样、长文本 30 字截断', () {
      final long = '题' * 34;
      final detail = jsonEncode({
        'fields': ['priority', 'status', 'title'],
        'changes': [
          {'field': 'priority', 'from': 0, 'to': 4},
          {'field': 'status', 'from': 'pending', 'to': 'doing'},
          {'field': 'title', 'from': '短标题', 'to': long},
        ],
      });
      expect(
        describeActivity('update', detail),
        '更新（优先级：无 → 紧急、状态：待办 → 进行中、标题：短标题 → ${'题' * 30}…）',
      );
    });

    test('旧行 fields：无 changes 回退字段名列表', () {
      expect(
        describeActivity('update', jsonEncode({
          'fields': ['due_date', 'priority']
        })),
        '更新（截止日期、优先级）',
      );
    });

    test('done/is_favorite/percent_done/position 值口径', () {
      String one(String f, Object? from, Object? to) =>
          describeActivity('update', jsonEncode({
            'changes': [
              {'field': f, 'from': from, 'to': to}
            ]
          }));
      expect(one('done', 0, 1), '更新（完成标记：未完成 → 已完成）');
      expect(one('is_favorite', 1, 0), '更新（收藏：是 → 否）');
      expect(one('percent_done', 0, 45), '更新（进度：0% → 45%）');
      expect(one('position', 1.2, 3.4), '更新（顺序：已调整 → 已调整）');
    });

    test('repeat_rule 伪字段：整规则快照走 repeatLabelExt', () {
      final snap = {
        'mode': 2,
        'after': 1,
        'weekdays': 0,
        'end_type': 0,
        'end_param': 0,
        'from_done': 0,
      };
      final weeklyMask = {...snap, 'weekdays': 0x11}; // bit0+bit4 = 周一+周五
      final weeklyFromDone = {...snap, 'from_done': 1};
      final detail = jsonEncode({
        'changes': [
          {'field': 'repeat_rule', 'from': {'mode': 0, 'after': 1, 'weekdays': 0, 'end_type': 0, 'end_param': 0, 'from_done': 0}, 'to': snap},
          {'field': 'repeat_rule', 'from': snap, 'to': weeklyMask},
          {'field': 'repeat_rule', 'from': snap, 'to': weeklyFromDone},
        ]
      });
      expect(
        describeActivity('update', detail),
        '更新（重复规则：不重复 → 每周、重复规则：每周 → 每周一五、'
        '重复规则：每周 → 每周（按完成日））',
      );
    });

    test('标签动作拼标签名', () {
      expect(
        describeActivity('label_add', jsonEncode({'label': '工作'})),
        '添加标签「工作」',
      );
      expect(
        describeActivity('label_remove', jsonEncode({'label': '工作'})),
        '移除标签「工作」',
      );
    });

    test('从属对象动作拼 target（子任务/评论/提醒/滚周期）', () {
      expect(
        describeActivity('subtask_add', jsonEncode({'target': '整理清单'})),
        '添加子任务「整理清单」',
      );
      expect(
        describeActivity('comment_add', jsonEncode({'target': '评审会周四开'})),
        '添加评论「评审会周四开」',
      );
      expect(
        describeActivity('repeat_rollover', jsonEncode({'target': '2026-09-26 09:00'})),
        '已滚动下一周期「2026-09-26 09:00」',
      );
    });

    test('create（含滚周期来源 detail）只报动作文案', () {
      expect(describeActivity('create', '{}'), '创建了任务');
      expect(
        describeActivity('create', jsonEncode({'from': 'repeat', 'parent_id': 7})),
        '创建了任务',
      );
    });

    test('detail 非法 JSON / 非对象 / 未映射 action 均回退', () {
      expect(describeActivity('update', 'not-json'), '更新');
      expect(describeActivity('update', '[1,2]'), '更新');
      expect(describeActivity('pinned', '{}'), 'pinned');
    });
  });
}
